codemod/output/AwsSignatureV4.ts

```diff
@@ -0,0 +1,199 @@
+import { AwsSignature } from "./AwsSignature";
+import { uri, awsUrl } from "./Utils";
+import { Global } from "./Global";
+
+class AwsSignatureV4 extends AwsSignature {
+	public _cr: any;
+	public payload: any;
+
+  constructor(request) {
+    super(request);
+    this._cr = undefined;
+  }
+
+  error() {
+    this._cr = undefined;
+  }
+
+  getPayload() {
+    return this.awsRequest.getPayload().then(data => {
+      this.payload = data;
+    });
+  }
+
+  authorizationString() {
+    const authParts = [];
+    const credentials = this.credentialString();
+    const headers = this.canonicalHeaders();
+    authParts.push(
+      ["AWS4-HMAC-SHA256 Credential=", this.con.aws_key, "/", credentials].join("")
+    );
+    authParts.push(`SignedHeaders=${headers.signedHeaders}`);
+    authParts.push(`Signature=${this.request.auth}`);
+    return authParts.join(", ");
+  }
+
+  stringToSign() {
+    const signParts = [];
+    signParts.push("AWS4-HMAC-SHA256");
+    signParts.push(this.request.dateString);
+    signParts.push(this.credentialString());
+    signParts.push(this.con.cryptoHexEncodedHash256(this.canonicalRequest()));
+    const result = signParts.join("\n");
+    Global.l.d("V4 stringToSign:", result);
+    return result;
+  }
+
+  credentialString() {
+    const credParts = [];
+    credParts.push(this.request.dateString.slice(0, 8));
+    credParts.push(this.con.awsRegion);
+    credParts.push("s3");
+    credParts.push("aws4_request");
+    return credParts.join("/");
+  }
+
+  canonicalQueryString() {
+    const qs = this.awsRequest.request.query_string || "";
+    const search = uri([this.awsRequest.awsUrl, this.request.path, qs].join(""))
+      .search;
+    const searchParts = search.length ? search.split("&") : [];
+    const encoded = [];
+    let nameValue;
+    let i;
+
+    for (i = 0; i < searchParts.length; i++) {
+      nameValue = searchParts[i].split("=");
+
+      encoded.push({
+        name: encodeURIComponent(nameValue[0]),
+        value: nameValue.length > 1 ? encodeURIComponent(nameValue[1]) : null
+      });
+    }
+
+    const sorted = encoded.sort((a, b) => {
+      if (a.name < b.name) {
+        return -1;
+      } else if (a.name > b.name) {
+        return 1;
+      }
+
+      return 0;
+    });
+
+    const result = [];
+
+    for (i = 0; i < sorted.length; i++) {
```

Comment:
Why not `for of`?

Reply:
Both are ugly if you ask me, but let's keep this PR purely a "codemod" thing and add additional refactorings later, what do you think?

As I wrote in #450 already, we should properly plan the "rewrite" of the next major version of Evaporate and split up the work.